### PR TITLE
apps/examples: add lwnl sample

### PR DESCRIPTION
**Pull request by pillip8282**
add lwnl sample. it creates 3 wi-fi event listener threads by lwnl and
it prints the type of event immediately.
1) run lwnl sample
   TASH> lwnl_sample
2) generate wi-fi manager event by wm_test
   TASH> wm_test start
   TASH> wm_test join ...
3) check that lwnl sample receive events
   [0] status (5) len(3268)␍␊
   [1] status (5) len(3268)␍␊
   [2] status (5) len(3268)

### CODE REVIEW

File: apps/examples/lwnl_sample/Make.defs [hunk 1]
```diff
@@ -0,0 +1,56 @@
+###########################################################################
+#
+# Copyright 2017 Samsung Electronics All Rights Reserved.
```
Comment (jeongchanKim) [inline apps/examples/lwnl_sample/Make.defs:3]: Please change to 2021.

File: apps/examples/lwnl_sample/Make.defs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+# either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+#
+###########################################################################
+############################################################################
+# apps/examples/hello/Make.defs
```
Comment (jeongchanKim) [inline apps/examples/lwnl_sample/Make.defs:19]: Please change the examples name.

File: apps/examples/lwnl_sample/Make.defs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+# OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+# AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+#
+############################################################################
+
+ifeq ($(CONFIG_EXAMPLES_HELLO),y)
```
Comment (jeongchanKim) [inline apps/examples/lwnl_sample/Make.defs:54]: Please change to `CONFIG_EXAMPLES_LWNL_SAMPLE`.

File: apps/examples/lwnl_sample/lwnl_sample.c [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+	int idx = *((int *)arg);
+	printf("[%d] run event listener\n", idx);
+	int fd = socket(AF_LWNL, SOCK_RAW, LWNL_ROUTE);
+	if (fd < 0) {
+		return NULL;
+	}
+
+	res = bind(fd, NULL, 0);
+	if (res < 0) {
+		printf("[%d] bind fail errno(%d)\n", idx, errno);
```
Comment (sangam-swami) [inline apps/examples/lwnl_sample/lwnl_sample.c:81]: can we close the fd is failure case and at the end of usage

File: apps/examples/lwnl_sample/lwnl_sample.c [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+	fd_set rfds, ofds;
+	FD_ZERO(&ofds);
+	FD_SET(fd, &ofds);
+
+	printf("[%d] run event loop\n", idx);
+	for (;;) {
+		rfds = ofds;
+		res = select(fd + 1, &rfds, NULL, NULL, 0);
+		if (res <= 0) {
+			printf("[%d] select error res(%d) errno(%d))\n", idx, res, errno);
```
Comment (sangam-swami) [inline apps/examples/lwnl_sample/lwnl_sample.c:94]: idx is already printed above, do we need to print it again here and down as well?

File: apps/examples/lwnl_sample/lwnl_sample.c [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+		if (FD_ISSET(fd, &rfds)) {
+			char buf[8] = {0, };
+			lwnl_cb_status status;
+			uint32_t len;
+			int nbytes = read(fd, (char *)buf, 8);
+			if (nbytes < 0) {
+				printf("[%d] read error bytes(%d) errno(%d))\n", idx, nbytes, errno);
+				return NULL;
+			}
```
Comment (dineshdas-samsung) [inline apps/examples/lwnl_sample/lwnl_sample.c:106]: Please add a extra line after the if construct  as per coding rule.

File: apps/examples/lwnl_sample/lwnl_sample.c [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+			memcpy(&status, buf, sizeof(lwnl_cb_status));
+			memcpy(&len, buf + sizeof(lwnl_cb_status), sizeof(uint32_t));
+			printf("[%d] status (%d) len(%d)\n", idx, status, len);
+
+			// flush the event queue
+			if (len > 0) {
+				char *tmp = (char *)malloc(len);
```
Comment (dineshdas-samsung) [inline apps/examples/lwnl_sample/lwnl_sample.c:113]: Please check  if tmp is null.

File: apps/examples/lwnl_sample/lwnl_sample.c [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+			return NULL;
+		}
+
+		if (FD_ISSET(fd, &rfds)) {
+			char buf[8] = {0, };
+			lwnl_cb_status status;
+			uint32_t len;
+			int nbytes = read(fd, (char *)buf, 8);
+			if (nbytes < 0) {
+				printf("[%d] read error bytes(%d) errno(%d))\n", idx, nbytes, errno);
```
Comment (dineshdas-samsung) [inline apps/examples/lwnl_sample/lwnl_sample.c:104]: Please make a note idx is also printed here.

File: apps/examples/lwnl_sample/lwnl_sample.c [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <net/if.h>
+#include <errno.h>
+#include <sys/select.h>
+#include <sys/socket.h>
+#include <tinyara/lwnl/lwnl.h>
+/*
+ * it's test code so it doesn't strictly error handling.
```
Comment (abhishek-samsung) [inline apps/examples/lwnl_sample/lwnl_sample.c:67]: Please change "strictly error handling." --> "strictly perform error handling." or "it's test code so it doesn't perform error handling strictly."

File: apps/examples/lwnl_sample/lwnl_sample.c [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+#ifdef CONFIG_BUILD_KERNEL
+int main(int argc, FAR char *argv[])
+#else
+int lwnl_sample_main(int argc, char *argv[])
+#endif
+{
+	pthread_t pid[3];
+
+	/*  run event listener thread */
+	for (int i = 0; i < 3; i++) {
```
Comment (davidfather) [inline apps/examples/lwnl_sample/lwnl_sample.c:130]: Can this test app have the number of test threads from a tester as an input parameter, not a fixed number ?
In this case, you should also add the help statement if the input parameter is missing.

File: apps/examples/lwnl_sample/lwnl_sample.c (hunk 5 above)
Comment (pillip8282) [reply apps/examples/lwnl_sample/lwnl_sample.c:94]: It's thread which shares same code. So printing idx is necessary.

File: apps/examples/lwnl_sample/lwnl_sample.c (hunk 6 above)
Comment (pillip8282) [reply apps/examples/lwnl_sample/lwnl_sample.c:106]: It doesn't violate the coding rule. Could you share the coding rule link?

File: apps/examples/lwnl_sample/lwnl_sample.c (hunk 7 above)
Comment (pillip8282) [reply apps/examples/lwnl_sample/lwnl_sample.c:113]: The variable tmp is valid inside if statement. So it doesn't need to be checked.

File: apps/examples/lwnl_sample/lwnl_sample.c (hunk 8 above)
Comment (pillip8282) [reply apps/examples/lwnl_sample/lwnl_sample.c:104]: Could you tell me more? I can't understand your comment.

**Review by davidfather — APPROVED** [review]
Looks good to me.

**Pull request merged**